ansibullbot/parsers/botmetadata.py

```diff
@@ -148,16 +149,35 @@ def _propagate(files, top, child, field, multivalued=True):
         def propagate_keys(data):
             files = data[u'files']
             '''maintainers and ignored keys defined at a directory level are copied to subpath'''
-            for file1, file2 in itertools.combinations(files.keys(), 2):
-                # Python 2.7 doesn't provide os.path.commonpath
-                common = os.path.commonprefix([file1, file2])
-                top = min(file1, file2)
-                child = max(file1, file2)
 
-                top_components = top.split(u'/')
-                child_components = child.split(u'/')
+            iterfiles = {}
+            filenames = sorted(list(files.keys()))
+            parent = None
+            for idp,parent in enumerate(filenames):
+
+                if parent not in iterfiles:
+                    iterfiles[parent] = []
+
+                started = False
+                for fn in filenames[idp:]:
+                    if parent == fn:
+                        continue
+                    if fn.startswith(parent):
+                        started = True
+                        iterfiles[parent].append(fn)
+                    elif started:
+                        break
+
+            for file1, files2 in iterfiles.items():
+                for file2 in files2:
+                    # Python 2.7 doesn't provide os.path.commonpath
+                    common = os.path.commonprefix([file1, file2])
+                    top = min(file1, file2)
+                    child = max(file1, file2)
+
+                    top_components = top.split(u'/')
+                    child_components = child.split(u'/')
```

Comment:
I think you can get rid of the common, top, child, top_components, and child_components setting now.  If I'm reading it right, these are all used to determine whether and which filename is the parent and which is the child.  That's now done when the information is put into iterfiles.  file1 == top and file2 == child.